File: uns/Practicos-Rust/ejercicio-2/java/main.java
import java.util.ArrayDeque;
import java.util.LinkedList;
import java.util.Queue;

class main {
    public static void soloImpares(Queue<Integer> cola) {
        // Creamos una cola temporal, donde guardaremos los valores impares.
        Queue<Integer> nuevaCola = new ArrayDeque<Integer>();
        while(!cola.isEmpty()) {

            // Comparamos el primer elemento de la cola, y en caso de ser impar, 
            // lo agregamos a nuestra cola temporal.
            int numeroAComparar = cola.remove();
            if((numeroAComparar % 2) == 1) {
                nuevaCola.add(numeroAComparar);
            }

        }
        
        // Volvemos a ingresar todos los valores impares a la cola original.
        while(!nuevaCola.isEmpty()) {
            cola.add(nuevaCola.remove());
        }
    }


    public static void main(String[] args) {
        // Genero la Cola usando la estructura ArrayDeque dada por java.
        Queue<Integer> myQueue = new ArrayDeque<Integer>();

        // agrego valores a la cola para probar mi funcion.
        for (int i = 1; i < 50; i += 3) {
            myQueue.add(i);
        }

        // Muestro en consola la Queue, aplico la función "soloImpares" y 
        // muestro el resultado por consola.
        System.out.println("Queue Completa:");
        for(int s : myQueue) { 
            System.out.println(s); 
        }
        
        soloImpares(myQueue);
        System.out.println("");
        System.out.println("Ahora solo los impares:");


        for(int s : myQueue) { 
            System.out.println(s); 
        }
    }

}

File: uns/Practicos-Rust/ejercicio-2/rust/main.rs
#[derive(Debug)]
struct Queue<T> {
    elementos: Vec<T>,
    top: i32,
    max: usize,
}

impl<T> Queue<T> {
    pub fn new(max: usize) -> Queue<T> {
        Queue {
            elementos: Vec::with_capacity(max),
            top: -1,
            max: max,
        }
    }

    pub fn add(&mut self, element: T) {
        if self.top < self.max as i32 {
            self.elementos.push(element);
            self.top += 1;
        }
    }

    pub fn remove(&mut self) -> T {
        self.top -= 1;
        self.elementos.pop().unwrap()
    }

    pub fn is_empty(&self) -> bool {
        if self.top == -1 {
            true
        } else {
            false
        }
    }

    pub fn length(&self) -> usize {
        self.max
    }
}

fn solo_impares(cola: &mut Queue<i32>) {
    let mut nueva_cola: Queue<i32> = Queue::new(cola.length());
    
    while !cola.is_empty() {
        let elemento = cola.remove();
        if elemento % 2 == 1 {
            nueva_cola.add(elemento);
        }
    }

    while !nueva_cola.is_empty() {
        cola.add(nueva_cola.remove())
    }
}

fn main() {
    let mut cola: Queue<i32> = Queue::new(10 as usize);

    for i in 1..=11 {
        cola.add(i);
    }

    println!("{:?}", cola);
    solo_impares(&mut cola);
    println!("{:?}", cola);
}